Using Robo Runner

SaraMcCutcheon opened the pull request:
Extracted from #1177

Commit:
Changed to using Robo's runner

dustinleblanc commented on src/Terminus.php:
```diff
@@ -2,47 +2,86 @@
+     * @return void
      */
-    protected function getDefaultInputDefinition()
+    private function setRunner()
```
A setter should probably explicitly accept an argument, it will be easier to test and also refactor later

dustinleblanc commented on src/Terminus.php:
```diff
@@ -2,47 +2,86 @@
-        $helper_set = parent::getDefaultHelperSet();
-        //$helper_set->set(new Pantheon\Terminus\Helpers\FileHelper());
-        return $helper_set;
+        $config = new Config();
```
Any reason we shouldn't inject the config object?

SaraMcCutcheon replied:
It's only used here and because we're switching to Robo's Runner, we can no longer inject it into our command classes.

dustinleblanc replied:
Cool! love that we don't have to maintain our own runner!

dustinleblanc commented on src/Commands/TerminusCommand.php:
```diff
@@ -52,11 +68,21 @@ private function ensureLogin() {
+     * @return void
+     */
+    private function setLog()
+    {
```
Should probably accept an argument of which logger to set

SaraMcCutcheon replied:
From the Symfony Console documentation, it appeared as though this process works automagically.
http://symfony.com/doc/current/components/console/logger.html
That is unless you mean that the argument should be an instance of LoggerInterface.

dustinleblanc replied:
Yeah, I'd think something like 

``` php
private function setLog(LoggerInterface $logger)
{
    $this->logger = new $logger($this->output());
    return $this;
}
```

calling of the method would then be

``` php
$logger = new FooLogger();
$this->setLog($logger);
```

greg-1-anderson replied:
Yes, just like that, but just $this->logger = $logger, no `new` here

dustinleblanc commented:
injecting those dependencies like a boss!

greg-1-anderson commented on src/Commands/TerminusCommand.php:
```diff
@@ -21,21 +24,34 @@
     {
-        $this->config = $config;
+        $this->config = new Config();
+        $this->setLog(new ConsoleLogger($this->output()));
```
Might want to inject a logger; I'd think you would definitely want to inject the Config() object, since setting that up probably involves parsing files.

SaraMcCutcheon replied:
How can I inject things into the commands Robo\Runner instantiates?

greg-1-anderson replied:
That's a reasonable request!  ;)

Robo is designed to use league/container inflection to automatically inject your dependencies.  Just make your class instantiate a SomethingAwareInterface, and the inflector will call the appropriate set method provided by SomethingAwareTrait in the same class.

However, you should not be REQUIRED to use league/container to use the Robo runner.  I'll make a small change to the runner to allow you to pass in already-constructed command classes.

greg-1-anderson replied:
Also, this line (`$this->setLog(new ConsoleLogger($this->output()));`) is the source of the verbosity bug.  What's happening here is that Robo injects the output object via the technique above ^^; it uses setOutput() in OutputAwareTrait, since you implement OutputAwareInterface.  The result of this is that $this->output() is not available in your constructor, as the constructor is called to set up the object before its setOutput() method is called.

n.b. if you construct your own objects, you'll need to inject the input and output objects yourself.

greg-1-anderson replied:
I should also mention that Robo sets up its own logger object, so if you just make TerminusCommand implement LoggerAwareInterface, then Robo will inject a working logger for you, available via `protected $logger;`, which is in LoggerAwareTrait.

greg-1-anderson replied:
use Psr\Log\LoggerAwareInterface;
use Psr\Log\LoggerAwareTrait;

SaraMcCutcheon replied:
Aha, THIS is just the thing I needed! WOOT!

greg-1-anderson commented on src/Terminus.php:
```diff
@@ -2,47 +2,90 @@
+    {
         $config = new Config();
-        parent::__construct('Terminus', $config->get('version'));
+        $status_code = $this->runner->execute($arguments, null, 'Terminus', $config->get('version'));
```
You should construct your own output object here, and pass it to Robo. Then, you'll have a reference to your output object, and can use it to construct your logger.

greg-1-anderson replied:
(Output object construction doesn't literally have to be right here; could pass it in, reference a field of this class, etc.)

greg-1-anderson commented on src/Terminus.php:
```diff
@@ -2,47 +2,90 @@
+          ),
+          function ($class_name) {
+              $reflection_class = new \ReflectionClass($class_name);
+              return !$reflection_class->isAbstract();
```
This is okay, but if you just put your abstract classes in a different namespace, then you wouldn't need to filter them out.

SaraMcCutcheon commented on src/Commands/TerminusCommand.php:
```diff
@@ -2,34 +2,36 @@
+    public function __construct()
     {
-        $this->config = $config;
+        $this->config = new Config();
```
Injecting this means that I won't be able to let Robo work its inputter/outputter/logger-injecting magic.

dustinleblanc replied:
Would using an ioc container make a difference? My brain starts to hurt when I think about service containers (and the yaml loading mess of Symfony's container makes it worse). If an upstream change in Robo would make it easier, we could submit a PR there too. For now, newing up an object in the constructor isn't the end of the world, but we should fight it like the plague before we take on too much tech debt.

greg-1-anderson replied:
It's really not that hard. Robo's injecting magic is designed to be extensible.
1. Make yourself a new TerminusConfigAwareInterface / Trait.
2. Make TerminusCommand implement the interface and use the trait.
3. Share your Config singleton in the container.
4. Add an inflector to the league/container at startup to set your config object on anything that implements TerminusConfigAwareInterface.

For that matter, you might be able to just get away with using Robo's config object. It has getters and setters; might work.  Or, subclass it and use Robo's ConfigAwareInterface / ConfigAwareTrait.

Yes, Robo rocks.  I'll show you how to do (3) and (4) tomorrow. It's just a few easy lines of code, because league/container rocks.  It's a pity that so many people try to learn DI with the Symfony container, because I agree that Symfony service containers are prone to making heads hurt. But it doesn't have to be painful.

SaraMcCutcheon commented on src/Runner.php:
```diff
@@ -2,87 +2,94 @@
-        $this->commands_directory = __DIR__ . '/Commands';
-        $this->application = new Terminus();
-        $this->configureApplication(new Config($options));
+        $this->application = $options['application'];
```
I would like to be able to use this application in Robo's Runner.

greg-1-anderson replied:
Good news:  that's easy!

$container = new \League\Container\Container();
$app = new Terminus(...);
$input = new \Symfony\Component\Console\Input\ArgvInput($argv);
$output = new \Symfony\Component\Console\Output\ConsoleOutput();
Robo::configureContainer($container, $input, $output, $app);

Then you can pass the container to the runner constructor. First parameter is still your commandfile list; the second parameter is null and the third parameter is the constructor.

greg-1-anderson replied:
I'm going to update this PR to do this right now.

greg-1-anderson replied:
Making good progress here; will have something to push shortly. I am going to check for abstract classes in annotated commands, though, so that we do not have to filter them here. I do think that you should move the TerminusCommand base class to a different namespace anyway, though.

dustinleblanc commented on src/Runner.php:
```diff
@@ -2,87 +2,98 @@
-            $subdirectories = array_merge($subdirectories, $this->locateSubdirectories($dir, $next_namespace));
-        }
-        return $subdirectories;
+        $commands = array_filter(
```
I <3 this functional waterfall of delight

greg-1-anderson replied:
I love it too -- even though I feel bad that you had to write it. ;) You can take it out when you update to annotated commands 1.3.1 and set the search locations to empty instead of ['.'].

greg-1-anderson commented on src/Runner.php:
```diff
@@ -2,87 +2,95 @@
-                }
-            }
-        }
+        $discovery->setSearchPattern('*Command.php')->setSearchLocations(['.']);
```
I'll give you a new CommandFileDiscovery that will work right if you change this to `setSearchLocations([]);`. The problem with `.` as a search location is it makes the command discovery object search each location twice, because it always searches the base directory atm.

greg-1-anderson replied:
I pushed a new consolidation/annotated-command version 1.3.1 that supports this.

greg-1-anderson committed:
Use Robo dependency injection to set up Symfony Style and Robo Logger objects. Demonstrate verbose/debug output with some temporary code in the Art command.

greg-1-anderson commented on src/Config.php:
```diff
@@ -228,6 +218,22 @@ private static function getTerminusScript()
     }
 
     /**
+     * Ensures that directory paths work in any system
```
I think this might blow up Cygwin, but I haven't used Cygwin in many years and don't quite remember.

dustinleblanc replied:
Hopefully with bash support in Win10 on it's way we will have less cause to care 😀

ari-gold replied:
I have Cygwin installed on a VirtualBox, so let me know if you'd like me to test.

greg-1-anderson commented:
I pushed a commit to improve DI here. I will also fix up the unit tests, but first I want to make the Robo API a bit better, so we can clean up things here a little more.

Commit:
Fixed missing-var bug, brushed up internal documentation

SaraMcCutcheon commented:
I fixed a couple bugs, brushed up the internal docs, and fixed the unit tests. Although there are still a few things that can be done to further enhance Terminus 1.0's use of Robo, I think this PR should be merged before we have the time to discuss those changes because it is blocking the work of others.

ronan commented:
This is looking sharp. I think any improvements will be minor and don't need to block this PR.

👍 👍 👍

bensheldon commented:
:+1: excited for the 🛫
